src/hmem_cuda_gdrcopy.c

```diff
@@ -0,0 +1,367 @@
+/*
+ * Copyright (c) 2020 Amazon.com, Inc. or its affiliates.
+ * All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "ofi_hmem.h"
+#include "ofi.h"
+
+#ifdef HAVE_GDRCOPY
+
+#include <pthread.h>
+#include <gdrapi.h>
+
+struct ofi_gdrcopy_handle {
+	gdr_mh_t mh; /* memory handler */
+	void *cuda_ptr; /* page aligned gpu pointer */
+	void *user_ptr; /* user space ptr mapped to GPU memory */
+	size_t length; /* page aligned length */
+};
+
+struct gdrcopy_ops {
+	gdr_t (*gdr_open)();
+	int (*gdr_close)(gdr_t g);
+	int (*gdr_pin_buffer)(gdr_t g, unsigned long addr, size_t size,
+			      uint64_t p2p_token, uint32_t va_space,
+			      gdr_mh_t *handle);
+	int (*gdr_unpin_buffer)(gdr_t g, gdr_mh_t handle);
+	int (*gdr_map)(gdr_t g, gdr_mh_t handle, void **va, size_t size);
+	int (*gdr_unmap)(gdr_t g, gdr_mh_t handle, void *va, size_t size);
+	int (*gdr_copy_to_mapping)(gdr_mh_t handle, void *map_d_ptr,
+				   const void *h_ptr, size_t size);
+	int (*gdr_copy_from_mapping)(gdr_mh_t handle, void *map_d_ptr,
+				     const void *h_ptr, size_t size);
+};
+
+static gdr_t global_gdr;
+pthread_spinlock_t global_gdr_lock;
```

Comment:
static pthread_spinlock_t global_gdr_lock;

Reply:
Thanks! fixed in new update.